src/librustc_privacy/lib.rs

```diff
@@ -412,14 +412,27 @@ impl<'b, 'a, 'tcx> ReachEverythingInTheInterfaceVisitor<'b, 'a, 'tcx> {
         self.ev.tcx.impl_trait_ref(self.item_def_id).visit_with(self);
         self
     }
+
+    fn visit_trait_ref(&mut self, trait_ref: ty::TraitRef<'tcx>) -> bool {
+        if let Some(node_id) = self.ev.tcx.hir.as_local_node_id(trait_ref.def_id) {
+            let item = self.ev.tcx.hir.expect_item(node_id);
+            self.ev.update(item.id, Some(AccessLevel::Reachable));
+        }
+
+        trait_ref.super_visit_with(self)
+    }
 }
 
 impl<'b, 'a, 'tcx> TypeVisitor<'tcx> for ReachEverythingInTheInterfaceVisitor<'b, 'a, 'tcx> {
     fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
         let ty_def_id = match ty.sty {
             ty::TyAdt(adt, _) => Some(adt.did),
             ty::TyDynamic(ref obj, ..) => obj.principal().map(|p| p.def_id()),
-            ty::TyProjection(ref proj) => Some(proj.trait_ref(self.ev.tcx).def_id),
+            ty::TyProjection(ref proj) => {
+                let trait_ref = proj.trait_ref(self.ev.tcx);
+                self.visit_trait_ref(trait_ref);
```

Comment:
Both `super_visit_ty` and `visit_trait_ref` will do the same thing (and the latter wasn't done before).